### ISSUE TITLE
Wishlist: MQTT plugin

### ISSUE THREAD

**Issue by njh** [collectd/collectd#805]
Hello,

I am considering writing an MQTT plugin for collectd. There has been some discussion on MQTT and collectd on the mailing list:
https://groups.google.com/d/msg/mqtt/bq8lisPrj8w/kMpAbSfY5GAJ

MQTT is a lightweight Pub/Sub protocol that was recently standardised by Oasis:
http://docs.oasis-open.org/mqtt/

If I submit a Pull Request, is it likely to be accepted?

Thanks!

nick.

**Comment by falzm** [comment]
Any updates on this plugin?

**Comment by mfournier** [comment]
@falzm I'm under the impression several concurrent implementations sprung to life a couple of months ago, but none really reached the point where it was fit for inclusion. I guess the authors were happy enough with an out-of-tree solution ?

That said, I personnally believe it would be a valuable plugin to have part of collectd, so if you're up continue from where one of them were left, I'd be glad to help ! (I'd favor something based on libmosquitto, because this lib is packaged in several major linux distros).

**Comment by falzm** [comment]
Hi @mfournier 

I too would prefer a native implementation similar to the AMQP plugin.

I'd could motivate myself to finish the plugin I've started (based on libmosquitto), but I was under the impression that @octo was taking over my original code to finish the plugin himself. If not, I could start from his branch and try to finish myself. WDYT @octo?

**Comment by jpmens** [comment]
I would very much like to see MQTT as a libmosquitto-based network plugin to be able to get collectd values in/out of servers, and I am *not* satisfied with an out-of-tree solution. At the time I worked a little on @octo 's tree but had to give up due to time constraints.

FWIW, I'd gladly help with testing as soon as we have something to test. :-)

**Comment by mfournier** [comment]
Implemented in #1124. This new plugin will be part of collectd 5.6.

Extra bonus if someone gives it a try before the release (by using the [nightly builds](https://github.com/collectd/collectd-ci/blob/master/README.md) for example), to ensure it works as expected, doesn't have any bugs, etc ;-)

**Comment by njh** [comment]
I have created a wiki page:
https://collectd.org/wiki/index.php/Plugin:MQTT

(updates/corrections to follow)

**Comment by njh** [comment]
I think there may be a problem with race condition while connecting:

```
~/Projects/collectd(master) $ collectd -f
[2015-09-07 18:32:08] [error] mqtt plugin: Unknown config option: CleanSession
[2015-09-07 18:32:08] [info] Initialization complete, entering read-loop.
[2015-09-07 18:32:08] [error] plugin mqtt: mosquitto_publish failed: The client is not currently connected.
[2015-09-07 18:32:08] [error] plugin mqtt: mosquitto_publish failed: The client is not currently connected.
[2015-09-07 18:32:08] [info] mqtt plugin: successfully reconnected to broker "localhost:1883"
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] plugin mqtt: mosquitto_publish failed: Broken pipe
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [error] mqtt plugin: publish failed: Unknown error.
[2015-09-07 18:32:08] [info] mqtt plugin: successfully reconnected to broker "localhost:1883"
```

Mosquitto logs:
```
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: Client mc-s097611 disconnected.
1441646894: New client connected from ::1 as mc-s097611 (c0, k60).
1441646894: Sending CONNACK to mc-s097611 (0, 0)
1441646894: Client mc-s097611 already connected, closing old connection.
1441646894: Client mc-s097611 disconnected.
1441646894: New client connected from ::1 as mc-s097611 (c0, k60).
1441646894: Sending CONNACK to mc-s097611 (0, 0)
1441646894: Client mc-s097611 already connected, closing old connection.
1441646894: Client mc-s097611 disconnected.
1441646894: New client connected from ::1 as mc-s097611 (c0, k60).
1441646894: Sending CONNACK to mc-s097611 (0, 0)
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: New connection from ::1 on port 1883.
1441646894: Client mc-s097611 already connected, closing old connection.
1441646894: Client mc-s097611 disconnected.
1441646894: New client connected from ::1 as mc-s097611 (c0, k60).
1441646894: Sending CONNACK to mc-s097611 (0, 0)
1441646894: Client mc-s097611 already connected, closing old connection.
1441646894: Client mc-s097611 disconnected.
1441646894: New client connected from ::1 as mc-s097611 (c0, k60).
1441646894: Sending CONNACK to mc-s097611 (0, 0)
1441646894: Received PUBLISH from mc-s097611 (d0, q0, r1, m0, 'collectd/mc-s097611/interface-stf0/if_octets', ... (19 bytes))
1441646894: Sending PUBLISH to mosqsub/82416-MC-S09761 (d0, q0, r0, m0, 'collectd/mc-s097611/interface-stf0/if_octets', ... (19 bytes))
```

Should I create a new GitHub issue?

**Comment by mfournier** [comment]
@njh, yes please, so were sure not to lose track of this problem. Thanks for having tested & given feedback!

**Comment by njh** [comment]
Created new issue #1262

**Comment by njh** [comment]
Created PR to fix subscribing: #1355
